source/code/plugin/filter_cadvisor2mdm.rb

```diff
@@ -0,0 +1,214 @@
+
+            # initialize cpu and memory limit 
+            if @process_incoming_stream
+                @cpu_limit = 0.0
+                @memory_limit = 0.0 
+                
+                begin 
+                    nodeInventory = JSON.parse(KubernetesApiClient.getKubeResourceInfo('nodes').body)
+                rescue Exception => e
+                    @log.info "Error when getting nodeInventory from kube API. Exception: #{e.class} Message: #{e.message} "
+                    ApplicationInsightsUtility.sendExceptionTelemetry(e.backtrace)
+                end
+                if !nodeInventory.nil? 
+                    cpu_limit_json = KubernetesApiClient.parseNodeLimits(nodeInventory, "capacity", "cpu", "cpuCapacityNanoCores")
+                    if !cpu_limit_json.nil? 
+                        @cpu_limit = cpu_limit_json[0]['DataItems'][0]['Collections'][0]['Value']
+                        @log.info "CPU Limit #{@cpu_limit}"
+                    else
+                        @log.info "Error getting cpu_limit"
+                    end
+                    memory_limit_json = KubernetesApiClient.parseNodeLimits(nodeInventory, "capacity", "memory", "memoryCapacityBytes")
+                    if !memory_limit_json.nil?
+                        @memory_limit = memory_limit_json[0]['DataItems'][0]['Collections'][0]['Value']
+                        @log.info "Memory Limit #{@memory_limit}"
+                    else
+                        @log.info "Error getting memory_limit"
+                    end
+                end
+            end
+        end
+
+        def build_metrics_hash
+            @log.debug "Building Hash of Metrics to Collect"
+            metrics_to_collect_arr = @metrics_to_collect.split(',').map(&:strip)
+            metrics_hash = metrics_to_collect_arr.map {|x| [x.downcase,true]}.to_h
+            @log.info "Metrics Collected : #{metrics_hash}"
+            return metrics_hash
+        end
+
+        def check_custom_metrics_availability
+            aks_region = ENV['AKS_REGION']
+            if aks_region.to_s.empty?
+                false # This will also take care of AKS-Engine Scenario. AKS_REGION is not set for AKS-Engine. Only ACS_RESOURCE_NAME is set
+            end
+            @log.debug "AKS_REGION #{aks_region}"
+            custom_metrics_regions_arr = @custom_metrics_azure_regions.split(',')
+            custom_metrics_regions_hash = custom_metrics_regions_arr.map {|x| [x.downcase,true]}.to_h
+
+            @log.debug "Custom Metrics Regions Hash #{custom_metrics_regions_hash}"
+
+            if custom_metrics_regions_hash.key?(aks_region.downcase)
+                @log.debug "Returning true for check_custom_metrics_availability"
+                true
+            else 
+                @log.debug "Returning false for check_custom_metrics_availability"
+                false
+            end
+        end
+
+		def shutdown
+			super
+		end
+
+        def filter(tag, time, record)
```

Comment:
Please add global rescue for all 'implicitly' called methods , else if you crash once, fluentD will unload your filter plugin and will not be running until the agent (and hence fluentD) is restarted.

Reply:
I will handle all HTTP Server Exceptions inside the filter, everything else will be handled by the global handler, else we can't do retry